### PR TITLE
Add OpenAI text-embedding-ada-002 API-based encoding

### PR DESCRIPTION
**Pull request by jasper-xian**
DEPENDENCIES ADDED: 
```
openai             0.28.0
tiktoken           0.4.0
```

TODO: incorporate OpenAI API query encoder into `pyserini.search`

To reproduce the [nfcorpus](https://github.com/castorini/pyserini/blob/master/docs/experiments-nfcorpus.md) experiment:

Indexing: 
```bash
python -m pyserini.encode \
  input   --corpus collections/nfcorpus/corpus.jsonl \
          --fields title text \
  output  --embeddings indexes/faiss.nfcorpus.contriever-msmacro \
          --to-faiss \
  encoder --encoder text-embedding-ada-002 \
          --encoder-class openai-api \
          --use-openai \
          --rate-limit 3500 \
          --device cpu \
          --dimension 1536 \
          --fields title text
```

Query Encoding:
```bash
python pyserini/encode/query.py \
--topics collections/nfcorpus/queries.tsv \
--encoder openai-api \
--max-length 512 \
--output openai-api/embedding.pkl
```

Retrieval with encoded queries from prior:
```bash
python -m pyserini.search.faiss \
  --encoded-queries openai-api \
  --index indexes/faiss.nfcorpus.contriever-msmacro \
  --topics collections/nfcorpus/queries.tsv \
  --output runs/run.beir-contriever-msmarco.nfcorpus.txt \
  --batch 128 --threads 16 \
  --hits 1000
```

Retrieval whilst encoding on the fly:
```bash
python -m pyserini.search.faiss \
  --encoder text-embedding-ada-002 \
  --encoder-class openai-api \
  --tokenizer cl100k_base \
  --index indexes/faiss.nfcorpus.contriever-msmacro \
  --topics collections/nfcorpus/queries.tsv \
  --output runs/run.beir-contriever-msmarco.nfcorpus.txt \
  --max-length 512 \
  --batch 128 --threads 16 \
  --hits 1000
```

Results: 
```
ndcg_cut_10             all     0.3676
```

### CODE REVIEW

**Comment by lintool** [comment]
If we're going to strictly stick to camel case naming conventions, then `OpenAI` should be `OpenAi`... but that just looks ugly to me...

File: pyserini/encode/__main__.py [hunk 1]
```diff
@@ -133,15 +137,26 @@ def parse_args(parser, commands):
...
+                    'texts': batch_info['text'],
+                    'titles': batch_info['title'] if 'title' in args.encoder.fields else None,
+                    'expands': batch_info['expand'] if 'expand' in args.encoder.fields else None,
+                    'fp16': args.encoder.fp16,
+                    'max_length': args.encoder.max_length,
+                    'add_sep': args.encoder.add_sep,
+                }
+                embeddings = encoder.encode(**kwargs)
+                batch_info['vector'] = embeddings
+                embedding_writer.write(batch_info, args.input.fields)
```
Comment (ronakice) [inline pyserini/encode/__main__.py:162]: A lot of this code is duplicated and can be condensed down to just the `int(args.encoder.rate_limit // (60 / OPENAI_API_RETRY_DELAY))` being different from `args.encoder.batch_size` so maybe clean that up?

**Review by ronakice — CHANGES_REQUESTED** [review]
Just this change for now!

**Comment by ronakice** [comment]
@jasper-xian we should merge this?

**Comment by jasper-xian** [comment]
yeah this should be good to merge @ronakice

**Review by ronakice — APPROVED** [review]
LGTM! Merging

**Comment by lintool** [comment]
@jasper-xian can you send a PR to add dependencies to `requirements.txt`?

**Pull request merged**